### PR TITLE
JAMES-3498 Split event bus

### PR DESCRIPTION
**Pull request by chibenwa**
- Moves the event bus to a separated API maven package
 - Generify the API as it has nothing specific to 'mailbox'
 - MailboxListener -> MailboxEvents
 - Preserve FQDN of GenericGroup for backward compatibility (despite it being in the API package this maeans having `mailbox` as a FQDN part)
 - Move `mailbox/events/*` to `event-bus/*`
 - Intriduce a EventSerializer so that we don't need mailbox-* dependencies for testing.

### CODE REVIEW

File: event-bus/api/src/test/java/org/apache/james/events/EventBusConcurrentTestContract.java [hunk 1]
```diff
@@ -50,33 +44,33 @@
...
+            EventBusTestFixture.EventListenerCountingSuccessfulExecution countingListener3 = newCountingListener();
 
             eventBus().register(countingListener1, new EventBusTestFixture.GroupA());
             eventBus().register(countingListener2, new EventBusTestFixture.GroupB());
             eventBus().register(countingListener3, new EventBusTestFixture.GroupC());
             int totalGlobalRegistrations = 3; // GroupA + GroupB + GroupC
 
             ConcurrentTestRunner.builder()
-                .operation((threadNumber, operationNumber) -> eventBus().dispatch(EVENT, NO_KEYS).block())
+                .operation((threadNumber, operationNumber) -> eventBus().dispatch(EventBusTestFixture.EVENT, EventBusTestFixture.NO_KEYS).block())
```
Comment (chibenwa) [inline event-bus/api/src/test/java/org/apache/james/events/EventBusConcurrentTestContract.java:73]: We need to get back static imports here and in other places of this file

File: event-bus/api/src/test/java/org/apache/james/events/GroupContract.java [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+            EventListener listener = event -> {
+                if (event.getEventId().equals(EVENT_ID)) {
+                    eventBus().dispatch(EventBusTestFixture.EVENT_2, EventBusTestFixture.NO_KEYS)
+                        .subscribeOn(Schedulers.elastic())
+                        .block();
+                    successfulRetry.set(true);
+                }
+            };
+
+            eventBus().register(listener, EventBusTestFixture.GROUP_A);
```
Comment (chibenwa) [inline event-bus/api/src/test/java/org/apache/james/events/GroupContract.java:157]: We need to get back static imports here and in other places of this file

File: event-bus/api/src/test/java/org/apache/james/events/KeyContract.java [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+            Mono.from(eventBus().register(event -> {
+                if (nbCalls.get() - finishedExecutions.get() > EventBus.EXECUTION_RATE) {
+                    rateExceeded.set(true);
+                }
+                nbCalls.incrementAndGet();
+                Thread.sleep(Duration.ofMillis(20).toMillis());
+                finishedExecutions.incrementAndGet();
+
+            }, EventBusTestFixture.KEY_1)).block();
```
Comment (chibenwa) [inline event-bus/api/src/test/java/org/apache/james/events/KeyContract.java:69]: We need to get back static imports here and in other places of this file

File: event-bus/cassandra/src/main/java/org/apache/james/events/CassandraEventDeadLettersDAO.java [hunk 4]
```diff
@@ -61,69 +57,69 @@
...
 
     private PreparedStatement prepareInsertStatement(Session session) {
-        return session.prepare(insertInto(TABLE_NAME)
-            .value(GROUP, bindMarker(GROUP))
-            .value(INSERTION_ID, bindMarker(INSERTION_ID))
-            .value(EVENT, bindMarker(EVENT)));
+        return session.prepare(insertInto(CassandraEventDeadLettersTable.TABLE_NAME)
+            .value(CassandraEventDeadLettersTable.GROUP, bindMarker(CassandraEventDeadLettersTable.GROUP))
+            .value(CassandraEventDeadLettersTable.INSERTION_ID, bindMarker(CassandraEventDeadLettersTable.INSERTION_ID))
+            .value(CassandraEventDeadLettersTable.EVENT, bindMarker(CassandraEventDeadLettersTable.EVENT)));
```
Comment (chibenwa) [inline event-bus/cassandra/src/main/java/org/apache/james/events/CassandraEventDeadLettersDAO.java:63]: We need to get back static imports here and in other places of this file

File: event-bus/cassandra/src/main/java/org/apache/james/events/CassandraEventDeadLettersGroupDAO.java [hunk 5]
```diff
@@ -49,22 +47,22 @@
...
     Mono<Void> storeGroup(Group group) {
         return executor.executeVoid(insertStatement.bind()
-                .setString(GROUP, group.asString()));
+                .setString(CassandraEventDeadLettersGroupTable.GROUP, group.asString()));
     }
 
     Flux<Group> retrieveAllGroups() {
         return executor.executeRows(selectAllStatement.bind())
-            .map(Throwing.function(row -> Group.deserialize(row.getString(GROUP))));
+            .map(Throwing.function(row -> Group.deserialize(row.getString(CassandraEventDeadLettersGroupTable.GROUP))));
```
Comment (chibenwa) [inline event-bus/cassandra/src/main/java/org/apache/james/events/CassandraEventDeadLettersGroupDAO.java:66]: We need to get back static imports here and in other places of this file

File: event-bus/cassandra/src/test/java/org/apache/james/events/CassandraEventDeadLettersDAOTest.java [hunk 6]
```diff
@@ -48,64 +36,64 @@
...
         cassandraEventDeadLettersDAO = new CassandraEventDeadLettersDAO(cassandraCluster.getConf(), eventSerializer);
     }
 
     @Test
     void removeEventShouldSucceededWhenRemoveStoredEvent() {
-        cassandraEventDeadLettersDAO.store(GROUP_A, EVENT_1, INSERTION_ID_1).block();
+        cassandraEventDeadLettersDAO.store(EventDeadLettersContract.GROUP_A, EventDeadLettersContract.EVENT_1, EventDeadLettersContract.INSERTION_ID_1).block();
 
-        cassandraEventDeadLettersDAO.removeEvent(GROUP_A, INSERTION_ID_1).block();
+        cassandraEventDeadLettersDAO.removeEvent(EventDeadLettersContract.GROUP_A, EventDeadLettersContract.INSERTION_ID_1).block();
```
Comment (chibenwa) [inline event-bus/cassandra/src/test/java/org/apache/james/events/CassandraEventDeadLettersDAOTest.java:47]: We need to get back static imports here and in other places of this file

File: event-bus/cassandra/src/test/java/org/apache/james/events/CassandraEventDeadLettersGroupDAOTest.java [hunk 7]
```diff
@@ -50,11 +48,11 @@ void retrieveAllGroupsShouldReturnEmptyWhenDefault() {
 
     @Test
     void retrieveAllGroupsShouldReturnStoredGroups() {
-        GROUP_DAO.storeGroup(GROUP_A).block();
-        GROUP_DAO.storeGroup(GROUP_B).block();
+        GROUP_DAO.storeGroup(EventDeadLettersContract.GROUP_A).block();
+        GROUP_DAO.storeGroup(EventDeadLettersContract.GROUP_B).block();
```
Comment (chibenwa) [inline event-bus/cassandra/src/test/java/org/apache/james/events/CassandraEventDeadLettersGroupDAOTest.java:52]: We need to get back static imports here and in other places of this file

File: event-bus/distributed/src/main/java/org/apache/james/events/GroupConsumerRetry.java [hunk 8]
```diff
@@ -53,7 +51,7 @@ static RetryExchangeName of(Group group) {
             return new RetryExchangeName(group.asString());
         }
 
-        static final String MAILBOX_EVENT_RETRY_EXCHANGE_PREFIX = MAILBOX_EVENT + "-retryExchange-";
+        static final String MAILBOX_EVENT_RETRY_EXCHANGE_PREFIX = RabbitMQEventBus.MAILBOX_EVENT + "-retryExchange-";
```
Comment (chibenwa) [inline event-bus/distributed/src/main/java/org/apache/james/events/GroupConsumerRetry.java:54]: We need to get back static imports here and in other places of this file

File: event-bus/distributed/src/main/java/org/apache/james/events/GroupRegistration.java [hunk 9]
```diff
@@ -60,7 +56,7 @@ static WorkQueueName of(Group group) {
             return new WorkQueueName(group);
         }
 
-        static final String MAILBOX_EVENT_WORK_QUEUE_PREFIX = MAILBOX_EVENT + "-workQueue-";
+        static final String MAILBOX_EVENT_WORK_QUEUE_PREFIX = RabbitMQEventBus.MAILBOX_EVENT + "-workQueue-";
```
Comment (chibenwa) [inline event-bus/distributed/src/main/java/org/apache/james/events/GroupRegistration.java:59]: We need to get back static imports here and in other places of this file

File: event-bus/distributed/src/main/java/org/apache/james/events/KeyReconnectionHandler.java [hunk 10]
```diff
@@ -51,7 +49,7 @@ public KeyReconnectionHandler(EventBusId eventBusId) {
     public Publisher<Void> handleReconnection(Connection connection) {
         return Mono.fromRunnable(() -> {
             try (Channel channel = connection.createChannel()) {
-                channel.queueDeclare(EVENTBUS_QUEUE_NAME_PREFIX + eventBusId.asString(), DURABLE, !EXCLUSIVE, AUTO_DELETE, QUEUE_ARGUMENTS);
+                channel.queueDeclare(KeyRegistrationHandler.EVENTBUS_QUEUE_NAME_PREFIX + eventBusId.asString(), DURABLE, !EXCLUSIVE, AUTO_DELETE, KeyRegistrationHandler.QUEUE_ARGUMENTS);
```
Comment (chibenwa) [inline event-bus/distributed/src/main/java/org/apache/james/events/KeyReconnectionHandler.java:52]: We need to get back static imports here and in other places of this file

File: event-bus/in-vm/pom.xml [hunk 11]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline event-bus/in-vm/pom.xml:1]: We miss a license

File: event-bus/in-vm/src/test/java/org/apache/james/events/delivery/InVmEventDeliveryTest.java [hunk 12]
```diff
@@ -32,39 +28,38 @@
...
 
     @Nested
     class SynchronousListener {
 
         @Test
         void deliverShouldDeliverEvent() {
-            when(listener.getExecutionMode()).thenReturn(MailboxListener.ExecutionMode.SYNCHRONOUS);
-            inVmEventDelivery.deliver(listener, EVENT, DeliveryOption.none())
+            when(listener.getExecutionMode()).thenReturn(EventListener.ExecutionMode.SYNCHRONOUS);
+            inVmEventDelivery.deliver(listener, EventBusTestFixture.EVENT, EventDelivery.DeliveryOption.none())
```
Comment (chibenwa) [inline event-bus/in-vm/src/test/java/org/apache/james/events/delivery/InVmEventDeliveryTest.java:62]: We need to get back static imports here and in other places of this file

File: mailbox/api/src/main/java/org/apache/james/events/MailboxIdRegistrationKey.java [hunk 13]
```diff
@@ -17,7 +17,7 @@
  * under the License.                                           *
  ****************************************************************/
 
-package org.apache.james.mailbox.events;
+package org.apache.james.events;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/events/MailboxIdRegistrationKey.java:20]: Oups no need to change this FQDN

File: mailbox/api/src/main/java/org/apache/james/events/MessageMoveEvent.java [hunk 14]
```diff
@@ -16,7 +16,7 @@
  * specific language governing permissions and limitations      *
  * under the License.                                           *
  ****************************************************************/
-package org.apache.james.mailbox.events;
+package org.apache.james.events;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/events/MessageMoveEvent.java:19]: Oups no need to change this FQDN

File: mailbox/api/src/test/java/org/apache/james/events/MailboxIdRegistrationKeyTest.java [hunk 15]
```diff
@@ -17,7 +17,7 @@
  * under the License.                                           *
  ****************************************************************/
 
-package org.apache.james.mailbox.events;
+package org.apache.james.events;
```
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/events/MailboxIdRegistrationKeyTest.java:20]: Oups no need to change this FQDN

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageMovesTest.java [hunk 16]
```diff
@@ -19,7 +19,7 @@
 
 package org.apache.james.mailbox.model;
 
-import org.apache.james.mailbox.events.MessageMoveEvent;
+import org.apache.james.events.MessageMoveEvent;
```
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageMovesTest.java:22]: Oups no need to change this FQDN

File: mailbox/event/json/src/main/scala/org/apache/james/event/json/MailboxEventSerializer.scala [hunk 17]
```diff
@@ -40,36 +41,36 @@ import play.api.libs.json._
 import scala.jdk.CollectionConverters._
 
 private sealed trait Event {
-  def toJava: JavaEvent
+  def toJava: events.Event
```
Comment (chibenwa) [inline mailbox/event/json/src/main/scala/org/apache/james/event/json/MailboxEventSerializer.scala:44]: Oups that JavaEvent convenience import was lost by the IDE

File: mailbox/store/src/test/resources/documents/html-iso-8859-1.html [hunk 18]
```diff
@@ -26,6 +26,6 @@
 		<title>*|MC:SUBJECT|*</title>
 
     <body>
-       <p>"�" this is a simple HTML text "�"</p>
+       <p>"�" this is a simple HTML text "�"</p>
```
Comment (chibenwa) [inline mailbox/store/src/test/resources/documents/html-iso-8859-1.html:29]: OUPS (I am not sure where it did come from...)

**Comment by chibenwa** [comment]
`[51/249]` test-jar kind of problems...

**Comment by chibenwa** [comment]
`[53/249]` test-jar kind of problems...

**Comment by chibenwa** [comment]
Armf `[64/249]`

**Comment by Arsnael** [comment]
```
[69/249]
[...]
[2ead150695b4a1794e3b1ff481499f1c283a282b] [ERROR]   LuceneMessageSearchIndexTest>AbstractMessageSearchIndexTest.setUp:113->initializeMailboxManager:43 » NoClassDefFound
```
Lots of similar other errors in the mailbox lucene module

**Comment by Arsnael** [comment]
I tried to read this PR... It's too hard to follow. I get it's a WIP, but I think I will wait for it to be finished and its git history getting a cleanup first before getting back in it

**Comment by chibenwa** [comment]
> I tried to read this PR... It's too hard to follow. I get it's a WIP, but I think I will wait for it to be finished and its git history getting a cleanup first before getting back in it

I'm unsure how much I will be able to clean up the history to be honest...

**Comment by Arsnael** [comment]
> I'm unsure how much I will be able to clean up the history to be honest...

I guess... it's quite a heavy refactoring here. Will do with it then

**Comment by chibenwa** [comment]
`[71/249]`

**Comment by Arsnael** [comment]
Quite a big change here... No ticket?

**Comment by chibenwa** [comment]
`[INFO] Building Apache James :: Mailbox :: Tools :: Copier 3.6.0-SNAPSHOT [83/249]`

**Comment by chibenwa** [comment]
There's an ADR, and we will create the ticket

**Comment by chibenwa** [comment]
`[142/249]`

**Comment by chibenwa** [comment]
`[164/249]`

**Comment by chibenwa** [comment]
```
[41ebc58f4b1a9ec061845e1f5fd9f74a6943b880] [INFO] Apache James :: Server :: Web Admin :: JMAP ........ FAILURE [  5.156 s]
[41ebc58f4b1a9ec061845e1f5fd9f74a6943b880] [INFO] Apache James :: Server :: Web Admin :: mailbox ..... SKIPPED
```

`[174/249]`

**Comment by Arsnael** [comment]
```
[cdedd48fd7021fd3635d30e13809a8226c85faac] Wrapped by: reactor.core.Exceptions$ErrorCallbackNotImplemented: org.apache.http.ConnectionClosedException: Connection closed unexpectedly
```
non related error...

test this please

**Comment by chibenwa** [comment]
`[INFO] Building Apache James :: Server :: Mailets 3.6.0-SNAPSHOT      [186/249]`

Small steps by small steps

**Comment by chibenwa** [comment]
`[INFO] Building Apache James :: Server :: Web Admin :: Mailbox :: Deleted Message Vault 3.6.0-SNAPSHOT [189/249]`

**Comment by chibenwa** [comment]
Mvn tests ran but...

```

org.jenkinsci.plugins.workflow.cps.steps.ParallelStepException: Parallel step 46d8aa09679e76e6fca1ee845236f9c7ae1bb6e7 failed
	at org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler$Callback.checkAllDone(ParallelStep.java:153)
	at org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler$Callback.onFailure(ParallelStep.java:119)
	at org.jenkinsci.plugins.workflow.cps.CpsBodyExecution$FailureAdapter.receive(CpsBodyExecution.java:294)
	at com.cloudbees.groovy.cps.impl.ThrowBlock$1.receive(ThrowBlock.java:68)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:58)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:145)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:19)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:33)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:106)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:164)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:274)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$000(CpsThreadGroup.java:74)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:183)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:181)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:47)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.ConnectException: Connection timed out
```

Quite obscure to me hence test this please

**Comment by chibenwa** [comment]
test this please

File: event-bus/api/src/test/java/org/apache/james/events/EventBusTestFixture.java [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(eventId, username);
+        }
+    }
+
+    class UnsupportedEventEvent implements Event {
```
Comment (Arsnael) [inline event-bus/api/src/test/java/org/apache/james/events/EventBusTestFixture.java:135]: `UnsupportedEvent` ?

File: mailbox/api/src/test/java/org/apache/james/mailbox/util/EventCollector.java [hunk 20]
```diff
@@ -22,11 +22,11 @@
...
 
-import org.apache.james.mailbox.events.Event;
-import org.apache.james.mailbox.events.Group;
-import org.apache.james.mailbox.events.MailboxListener;
+import org.apache.james.events.Event;
+import org.apache.james.events.EventListener;
+import org.apache.james.events.Group;
 
-public class EventCollector implements MailboxListener.GroupMailboxListener {
+public class EventCollector implements EventListener.GroupEventListener {
```
Comment (Arsnael) [inline mailbox/api/src/test/java/org/apache/james/mailbox/util/EventCollector.java:29]: Didn't you recreate this class in events module? it's a duplicate then and should be maybe removed here?

Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/util/EventCollector.java:29]: > Didn't you recreate this class in events module? 

Yes

> it's a duplicate then and should be maybe removed here?

No. I wanted to avoid a painful testing dependency.

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged
